src/main/java/edu/cornell/kfs/fp/document/web/struts/RecurringDisbursementVoucherAction.java

```diff
@@ -61,6 +68,44 @@ public ActionForward copy(ActionMapping mapping, ActionForm form, HttpServletReq
     	recurringForm.getPdpStatuses().clear();
     	ActionForward forward = super.copy(mapping, form, request, response);
         return forward;
+        
+    }
+    
+    public ActionForward confirmAndCancel(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        RecurringDisbursementVoucherForm recurringForm = (RecurringDisbursementVoucherForm) form;
+        RecurringDisbursementVoucherDocument recurringDV = (RecurringDisbursementVoucherDocument) recurringForm.getDocument();
+        
+        if (StringUtils.isEmpty(recurringDV.getPaymentCancelReason())) {
+            String paymentCancelFieldName = AccountingDocumentRuleBaseConstants.ERROR_PATH.DOCUMENT_ERROR_PREFIX + 
+                    CUKFSPropertyConstants.RECURRING_DV_PAYMENT_CANCEL_REASON_FIELD_NAME;
+            GlobalVariables.getMessageMap().putError(paymentCancelFieldName, CUKFSKeyConstants.ERROR_RCDV_PAYMENT_CANCEL_REASON_REQUIRED);
+        } else {
+            Set<String> canceledPaymentGroups = getRecurringDisbursementVoucherDocumentService().cancelPDPPayments(recurringDV, recurringDV.getPaymentCancelReason());
+            Set<String> canceledDVs = getRecurringDisbursementVoucherDocumentService().cancelDisbursementVouchers(recurringDV, recurringDV.getPaymentCancelReason());
+            reportCanceledItemsToBrowser(canceledPaymentGroups, canceledDVs);
+        }
+        forceRefreshOfPDPStatuses(recurringForm);
+        return mapping.findForward(RiceConstants.MAPPING_BASIC);
+    }
+    
+    private void reportCanceledItemsToBrowser (Set<String> canceledPaymentGroups, Set<String> canceledDVs) {
+       if (isSetPopulated(canceledPaymentGroups)) {
+           KNSGlobalVariables.getMessageList().add(CUKFSKeyConstants.INFO_RCDV_LIST_OF_PAYMENT_GROUPS_CANCELED, StringUtils.join(canceledPaymentGroups, ", "));
+       }
+       if (isSetPopulated(canceledDVs)) {
+           KNSGlobalVariables.getMessageList().add(CUKFSKeyConstants.INFO_RCDV_LIST_OF_DVS_CANCELED, StringUtils.join(canceledDVs, ", "));
+       }
+       if (!isSetPopulated(canceledPaymentGroups) && !isSetPopulated(canceledDVs)) {
+           KNSGlobalVariables.getMessageList().add(CUKFSKeyConstants.INFO_RCDV_NO_PAYMENTS_OR_DVS_CANCELED);
+       }
+    }
+    
+    private boolean isSetPopulated(Set<String> setToTest) {
+        return setToTest != null && !setToTest.isEmpty();
```

Comment:
If the methods on RecurringDisbursementVoucherDocumentService are not expected to return a null Set, then you can probably remove this method and replace its usage with isEmpty() calls on the Sets instead. (Generally, I think we want our Kuali services to avoid returning null collections or maps.)